opentech/apply/projects/views.py

```diff
@@ -11,20 +12,81 @@
 from opentech.apply.activity.messaging import MESSAGES, messenger
 from opentech.apply.activity.views import ActivityContextMixin, CommentFormView
 from opentech.apply.users.decorators import staff_required
-from opentech.apply.utils.views import (DelegateableView, DelegatedViewMixin,
-                                        ViewDispatcher)
+from opentech.apply.utils.views import (
+    DelegateableView,
+    DelegatedViewMixin,
+    ViewDispatcher
+)
 
 from .forms import (
+    ApproveContractForm,
     CreateApprovalForm,
     ProjectApprovalForm,
     ProjectEditForm,
     RejectionForm,
     RemoveDocumentForm,
     SetPendingForm,
     UpdateProjectLeadForm,
-    UploadDocumentForm,
+    UploadContractForm,
+    UploadDocumentForm
 )
-from .models import CONTRACTING, Approval, Project, PacketFile
+from .models import CONTRACTING, Approval, Contract, PacketFile, Project
+
+
+class ContractsMixin:
+    def get_context_data(self, **kwargs):
+        project = self.get_object()
+        recent_contracts = list(project.contracts.filter(is_signed=True)
+                                                 .order_by('-created_at')[:5])
+
+        # display an unsigned contract if it's more recent than the signed ones
+        latest_unsigned = project.contracts.filter(is_signed=False).first()
```

Comment:
```python
contracts = project.contracts.order_by('-created_at')
latest_contract = contracts.select_related('approver').first()
signed_contracts = contracts.filter(is_signed=True)
if latest_contract and latest_contract.approver:
    signed_contracts = signed_contracts.exclude(pk=latest_contract.pk)
```

This would change the FE code a bit, but would also allow us to insert the approve button closed to the contract that needs approving.

Reply:
We could insert the approve button without changing the view like so:

```
{% if forloop.first and not contract.approver %}<button>{% endif %}
```

Which do you prefer?

Reply:
I took your idea and expanded on it a bit in e0172c3b, adding tests as this turned out to be kind of tricky.